### DISCUSSION TITLE
How to use the plugin

### DISCUSSION THREAD

**Discussion by MathieuBujold** [spatie/laravel-navigation#12] [Q&A] [answered]
Hi
I'm trying to make the plugins to work on my project in Laravel 9.1.0

So far I did :
1. Install the plugin with composer
2. add the plugin to the config/app.php provider
       App\Providers\NavigationProvider::class,
3. Add the alias in the config/app.php aliases
      'Navigation' => Spatie\Navigation\NavigationServiceProvider::class,
4. From what I understand I create a new provider
      php artisan make:provider NavigationProvider
5. Inside the provider I add to the top 
      use Spatie\Navigation\Navigation;
6. Inside the handle function I add this
      app(Navigation::class)
             ->add('dashboard', route('dashboard'));

I have the following error since:
       Route [dashboard] not defined.

In my routes/web.php I have the following route.
        Route::get('/', function () {
            return view('pages.dashboard');
        })->name('dashboard');

Any Idea what I miss.
Any tips also on how I will use this in the blade after making the route portion work.

Thank you for your help.

**Comment by MathieuBujold** [comment]
I figure how to make the plugin work.
I'm not sure it's the most elegant way to do this.
Let me know your comment.

Inside the boot method of my controller I did this.

`
$this->app->booted(function () {
            app(Navigation::class)
            ->add('dashboard', route('dashboard'));
});
view()->composer('*',function($view) {
            $view->with('navigation', app(Navigation::class)->tree() );
            $view->with('breadcrumbs', app(Navigation::class)->breadcrumbs() );
});`

**Comment by MathieuBujold** [comment]
An other question.
In the service provider that I create, I can't access Auth It's always null.
I'm not sure the best place is to put this in a Provider
Any idea.

**Reply by freekmurze** [reply]
You probably are putting in the `register` method. Try using the `boot` method, which will be called later in the request cycle.

**Reply by MathieuBujold** [reply]
Thank you for your reply
It is in the boot but Auth:: return me null
The only way I figure to make this to work is this inside the boot method then inside a view composer 

 view()->composer('layout.shared.leftnav' , function($view){

}

At the end of the view composer I return the value with the $view->with

What do you think of this fix.

**Comment by jameswagoner** [comment, marked as answer]
You are correct that have to build your navigation inside of `$this->app->booted()` callback if you want to to use a service provider.

Alternatively, you could build your navigation in a middleware.  This was my approach since I am using Inertia and used the `HandleInertiaRequests` middleware.

```php
public function share(Request $request)
{
    return array_merge(parent::share($request), [
        'navigation' => $this->buildNavigation()->tree()
    ]
}

protected function buildNavigation()
{
    return app(Navigation::class)->addIf(
        auth()->user()->can('manage'),
        'Users',
        route('users')
    );
}
```
